marshmallow_dataclass/__init__.py

```diff
@@ -81,88 +82,192 @@ class User:
 # Recursion guard for class_schema()
 _RECURSION_GUARD = threading.local()
 
+if sys.version_info >= (3, 10):
+    # match_args, kw_only, slots
+
+    @overload
+    def dataclass(
+        _cls: Type[_U],
+        *,
+        repr: bool = True,
+        eq: bool = True,
+        order: bool = False,
+        unsafe_hash: bool = False,
+        frozen: bool = False,
+        match_args: bool = True,
+        kw_only: bool = False,
+        slots: bool = False,
+        base_schema: Optional[Type[marshmallow.Schema]] = None,
+        cls_frame: Optional[types.FrameType] = None,
+    ) -> Type[_U]:
+        ...
 
-@overload
-def dataclass(
-    _cls: Type[_U],
-    *,
-    repr: bool = True,
-    eq: bool = True,
-    order: bool = False,
-    unsafe_hash: bool = False,
-    frozen: bool = False,
-    base_schema: Optional[Type[marshmallow.Schema]] = None,
-    cls_frame: Optional[types.FrameType] = None,
-) -> Type[_U]:
-    ...
-
-
-@overload
-def dataclass(
-    *,
-    repr: bool = True,
-    eq: bool = True,
-    order: bool = False,
-    unsafe_hash: bool = False,
-    frozen: bool = False,
-    base_schema: Optional[Type[marshmallow.Schema]] = None,
-    cls_frame: Optional[types.FrameType] = None,
-) -> Callable[[Type[_U]], Type[_U]]:
-    ...
-
+    @overload
+    def dataclass(
+        *,
+        repr: bool = True,
+        eq: bool = True,
+        order: bool = False,
+        unsafe_hash: bool = False,
+        frozen: bool = False,
+        match_args: bool = True,
+        kw_only: bool = False,
+        slots: bool = False,
+        base_schema: Optional[Type[marshmallow.Schema]] = None,
+        cls_frame: Optional[types.FrameType] = None,
+    ) -> Callable[[Type[_U]], Type[_U]]:
+        ...
 
-# _cls should never be specified by keyword, so start it with an
-# underscore.  The presence of _cls is used to detect if this
-# decorator is being called with parameters or not.
-def dataclass(
-    _cls: Type[_U] = None,
-    *,
-    repr: bool = True,
-    eq: bool = True,
-    order: bool = False,
-    unsafe_hash: bool = False,
-    frozen: bool = False,
-    base_schema: Optional[Type[marshmallow.Schema]] = None,
-    cls_frame: Optional[types.FrameType] = None,
-) -> Union[Type[_U], Callable[[Type[_U]], Type[_U]]]:
-    """
-    This decorator does the same as dataclasses.dataclass, but also applies :func:`add_schema`.
-    It adds a `.Schema` attribute to the class object
+    # _cls should never be specified by keyword, so start it with an
+    # underscore.  The presence of _cls is used to detect if this
+    # decorator is being called with parameters or not.
+    def dataclass(
+        _cls: Type[_U] = None,
```

Comment:
Implicit optional parameter types are deprecated[^1].  Should fix to `__cls: Optional[Type[_U]] = None` while we're in here.  (Though there are admittedly a lot of other places in our code where we rely on implicit optional...)

[^1]: https://mypy-lang.blogspot.com/2022/09/mypy-0981-released.html